posts/2023-11-03_higher_order/higher_order.qmd

```diff
@@ -0,0 +1,290 @@
+
+`slice_derivation()` comes to the rescue! 
+
+* Once again, the function to restrict, is passed through the `derivation` argument;
+* The arguments that remain constant across slides are passed in the `args` selection using a call to `params()`;
+* The user passes `derivation_slice`'s to the function detailing the filter condition for the slice in the `filter` argument and what differs across runs in the `args` call. 
+
+```{r}
+adae_ahsev2fl <- adae %>%
+  slice_derivation(
+    derivation = derive_var_extreme_flag,
+    args = params(
+      new_var = AHSEV2FL,
+      by_vars = exprs(USUBJID)
+    ),
+    derivation_slice(
+      filter = AESTDY >= 1,
+      args = params(order = exprs(TEMP_AESEVN, AESTDY, AESEQ), mode = "first")
+    ),
+    derivation_slice(
+      filter = TRUE,
+      args = params(order = exprs(AESTDY, AESEQ), mode = "last")
+    )
+  )
+```
+
+```{r, eval=TRUE, echo=FALSE}
+adae_ahsev2fl %>%
+  arrange(USUBJID, AESTDY, AESEQ, desc(TEMP_AESEVN)) %>%
+  dataset_vignette(
+    display_vars = exprs(USUBJID, AEDECOD, AESTDY, AESEQ, AESEV, AHSEV2FL)
+  )
+```
+
+Notice that the `derivation_slice` ordering is important. in the above examples, all the AEs on or after study day 1 were addressed first, and then the `filter = TRUE` option was employed to catch all remaining records (in this case pre-treatment AEs).
+
+The ordering is perhaps shown even more when in the below example where three slices are taken. Remember that observations that match with more than one slice are only considered for the first matching slice. Thus, in this case the objective is to create a flag for each patient for the record with the first severe AE, and then the first moderate AE, and finally the last occurring AE which is neither severe or moderate.
+
+```{r}
+adae_ahsev3fl <- adae %>%
+  slice_derivation(
+    derivation = derive_var_extreme_flag,
+    args = params(
+      new_var = AHSEV3FL,
+      by_vars = exprs(USUBJID)
+    ),
+    derivation_slice(
+      filter = AESEV == "SEVERE",
+      args = params(order = exprs(AESTDY, AESEQ), mode = "first")
+    ),
+    derivation_slice(
+      filter = AESEV == "MODERATE",
+      args = params(order = exprs(AESTDY, AESEQ), mode = "first")
+    ),
+    derivation_slice(
+      filter = TRUE,
+      args = params(order = exprs(AESTDY, AESEQ), mode = "last")
+    )
+  )
+```
+
+```{r, eval=TRUE, echo=FALSE}
+adae_ahsev3fl %>%
+  arrange(USUBJID, AESTDY, AESEQ) %>%
+  dataset_vignette(
+    display_vars = exprs(USUBJID, AEDECOD, AESTDY, AESEQ, AESEV, AHSEV3FL)
+  )
+```
+
+The order is only important when the slices are not mutually exclusive, so in the above case the moderate AE slice could have been above the severe AE slice, for example, and there would have been no difference to the result. However the third slice had to come last to check all remaining (i.e. not severe or moderate) records only.
```

Comment:
Do I have this right: 

the slicing function will never override previous slices if the filters overlap?

Reply:
Maybe you could clarify the hierarchy of computation?

Reply:
Maybe clarify is not the right word, but to make it a bit more explicit

Reply:
clarified further up in the article